### ISSUE TITLE
New Check: IfCanBeSwitch 

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#15226]
child of #14961

In many cases, a series of if-else statements can be simplified using a switch statement, especially when dealing with multiple branches. Using a switch improves code readability and maintainability.

This new check will report cases where an if-else chain can be converted to a switch statement. The check includes a property to specify the minimum number of if-else branches required for the check to trigger. 

Note that this check only applies when each branch is checking the same variable, not when different variables are checked, because in this case it can't be converted to a switch. Switch should only have one selector expression that is being checked on multiple values

So we should violate the if-else chain only when the same variable is being checked in every branch

---

**Config**

$ cat config.xml 
```XML
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="IfCanBeSwitch">
            <property name="minimumIfBranches" value="3"/>
        </module>
    </module>
</module>
```
---

**Examples**

```Java
void test(String str) {  // violation, If-else chain can be converted to a switch statement
    if (str.equals("1")) {  // 1
        System.out.println(1);
    } else if (str.equals("2")) {  // 2
        System.out.println(2);
    } else if (str.equals("3")) {  // 3
        System.out.println(3);
    } else {
        System.out.println(4);
    }
}

void test2(String str) {  // ok, does not meet the minimum branch requirement, (else is not counted)
    if (str.equals("1")) {  // 1
        System.out.println(1);
    } else if (str.equals("2")) {  // 2
        System.out.println(2);
    } else {
        System.out.println(2);
    }
}

void test3(String str, int num) { // ok, different variables are checked in each branch
    if (str.equals("1")) {
        System.out.println(1);
    } else if (str.equals("2")) {
        System.out.println(2);
    } else if (num == 3) {
        System.out.println(3);
    } else {
        System.out.println(4);
    }
}

void test4(String str, int num) { // ok, different variables  are checked in each branch
    if (str.equals("1") && num == 1) {
        System.out.println(1);
    } else if (str.equals("2") && num == 2) {
        System.out.println(2);
    } else if (num == 3 && num == 3) {
        System.out.println(3);
    } else {
        System.out.println(4);
    }
}
```

---
**Output**

```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml IfCanBeSwitchExample.java 
Starting audit...
[ERROR] D:\IfCanBeSwitchExample.java:3:5: If-else chain can be converted to a switch statement.
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by rnveach** [comment]
How are you going to determine if all the conditions of if are using the same variable, and this can be used in an if? Will this only handle the very basic examples?

Before Java X, strings couldn't be in switches. Before Java Y, you couldn't have patterns in switches. Variables can be in different ordering (`A.equals(B)` versus `B.equals(A)`, we are using equals on an non-constant (requirement of a case), same variables can be using different operators (`equals` versus `==` versus `>=`), or we are calling methods and not an operator.

I would need some more context on what you will be validating and what you want with some examples.

**Comment by rnveach** [comment]
> Switch was first introduced in [Java 7](https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.11). It comes with strings in the first release

You could do switches before Java 7 and they didn't support strings. ChatGPT said it has been there since 1.0 . I use to enforce Java 6 on my project at work.

> not all IDENT will be considered ofc, we will decide which one to check its text and which to ignore

You need to expand on this in the first post and update with more examples. We need a clear "what we will consider" and "what we won't consider".

**Comment by mahfouz72** [comment]
> You could do switches before Java 7 and they didn't support strings. ChatGPT said it has been there since 1.0 . I use to enforce Java 6 on my project at work.

we can mention that this check should be used for java 7+ ?

> You need to expand on this in the first post and update with more examples. We need a clear "what we will consider" and "what we won't consider".

I think u misunderstood me this sentence is AST-specific. I wanted to say that we should stop only on `IDENT` on the `if` branches not any `IDENT`

**Comment by rnveach** [comment]
I am assigning to @mahfouz72 . I don't really get a clear picture of the type of ifs we will be processing.

Examples only show `A.equals(B)`. Are we only looking for those specific types?

What about enumerations? Those look like variables and switches have to be constants, so I assume we will ignore them?

**Comment by nrmancuso** [comment]
I do not think that Checkstyle is capable of making this a valuable check, since we could only handle some of the most simple cases due to not being type aware, IMO.

**Comment by mahfouz72** [comment]
After thinking more on this I tend to agree if we want to make this check it will support very simple cases only. I think we can close this. This  check needs us to be type aware in many cases. 
@rnveach @nrmancuso please decide should we close this ?

**Comment by nrmancuso** [comment]
Let’s leave it open for posterity, maybe someone will come along with some new idea, or the language feature will change somehow to make this easier. I am removing it from the project.

**Comment by romani** [comment]
Better to close, we can always to reopen and we can always discuss even in closed issues.
